### PR TITLE
Add CouchDB Status Indicator

### LINKED ISSUE #4888: OpenMCT should tell you it cannot connect to its persistence store
**Is your feature request related to a problem? Please describe.**
If OpenMCT cannot connect to its persistence store (e.g., CouchDB), there's no notification to the user aside from some exciting errors in the console. The stored objects in CouchDB don't appear, and the first error a user will get is if they try to save something and the dialog hangs.

**Describe the solution you'd like**
We should alert the user that the configured persistence store is unreachable.

**Describe alternatives you've considered**
If they have more than one persistence store, perhaps one being down is acceptable? In that case, I think an operator would still like to know that one of them is unavailable.

**Additional context**
The error message if CouchDB is down:
```
CouchObjectProvider.js?9318:132          GET http://localhost:5984/openmct/mine net::ERR_CONNECTION_REFUSED
```
The frozen save dialog:
Screen Shot 2022-02-23 at 12 36 07 PM

### PR DESCRIPTION
**Pull request by ozyx**
Closes #4888 

### Describe your changes:

Adds a status indicator to the CouchDB plugin which informs whether the connection is active, inactive, or pending (attempting to reconnect).

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [x] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Review by scottbell — CHANGES_REQUESTED** [review]
Good work! I've got a few changes requests below if you please

File: src/plugins/persistence/couch/CouchChangesFeed.js [hunk 1]
```diff
@@ -61,10 +68,34 @@
...
+        case EventSource.CONNECTING:
+            message.state = 'pending';
+            break;
+        case EventSource.OPEN:
+            message.state = 'open';
+            break;
+        case EventSource.CLOSED:
+            message.state = 'close';
+            break;
+        }
```
Comment (scottbell) [inline src/plugins/persistence/couch/CouchChangesFeed.js:95]: We should have a default case here - and maybe throw an error if we get something strange?

File: src/plugins/persistence/couch/CouchObjectProvider.js [hunk 2]
```diff
@@ -103,6 +108,23 @@ class CouchObjectProvider {
         }
     }
 
+    _messageToIndicatorState(message) {
```
Comment (scottbell) [inline src/plugins/persistence/couch/CouchObjectProvider.js:111]: we should make this a private method with `#`

File: src/plugins/persistence/couch/CouchObjectProvider.js [hunk 3]
```diff
@@ -103,6 +108,23 @@ class CouchObjectProvider {
...
+        case 'open':
+            state = CONNECTED;
+            break;
+        case 'close':
+            state = DISCONNECTED;
+            break;
+        case 'pending':
+            state = PENDING;
+            break;
+        }
```
Comment (scottbell) [inline src/plugins/persistence/couch/CouchObjectProvider.js:123]: we should have a default case here

File: src/plugins/persistence/couch/CouchObjectProvider.js [hunk 4]
```diff
@@ -129,14 +151,20 @@ class CouchObjectProvider {
...
+            } else if (response.status === CouchObjectProvider.HTTP_NOT_FOUND
+                || response.status === CouchObjectProvider.HTTP_SERVER_ERROR) {
+                this.indicator.setIndicatorToState(MAINTENANCE);
+            }
 
-                return response.json();
-            });
+            return await response.json();
+        } catch (err) {
+            this.indicator.setIndicatorToState(DISCONNECTED);
```
Comment (scottbell) [inline src/plugins/persistence/couch/CouchObjectProvider.js:166]: We should probably log the error here too

**Review by scottbell — CHANGES_REQUESTED** [review]

**Comment by ozyx** [comment]
@scottbell I was able to add some tests to check the status changes triggered by successful / failed `request` calls within `CouchObjectProvider`, but I was not able to figure out how to test status changes triggered by `SharedWorker` event messages. 

For example, the `PENDING` status is only triggered if there is an active `SharedWorker` listening to the changes feed, an event fires, and the `EventSource` reports its `readyState` as 'connecting'. I'm not sure how best to simulate this.

File: src/plugins/persistence/couch/CouchObjectProvider.js (hunk 3 above)
Comment (scottbell) [reply src/plugins/persistence/couch/CouchObjectProvider.js:132]: agreed, though i think we'll need to work that through the team

**Comment by scottbell** [comment]
> @scottbell I was able to add some tests to check the status changes triggered by successful / failed `request` calls within `CouchObjectProvider`, but I was not able to figure out how to test status changes triggered by `SharedWorker` event messages.
> 
> For example, the `PENDING` status is only triggered if there is an active `SharedWorker` listening to the changes feed, an event fires, and the `EventSource` reports its `readyState` as 'connecting'. I'm not sure how best to simulate this.

Agree, in that case it's going to be pretty hard to simulate a full round trip. Maybe we should short circuit the worker/event sourcen logic in that case. We could just ensure `PENDING` is accepted by the indicator by manual calling `onSharedWorkerMessage` in the provider with a fake `state` message. What do you think?

**Comment by scottbell** [comment]
Fantastic work!

**Pull request merged**